employees/tests/test_unit_report_viewset.py

```diff
@@ -0,0 +1,143 @@
+from rest_framework.reverse import reverse
+from employees.models import Report
+from users.models import CustomUser
+from managers.models import Project
+import datetime
+from decimal import Decimal
+from employees.views import ReportViewSet
+from rest_framework.test import APIRequestFactory
+from django.contrib.auth.models import AnonymousUser
+
+
+class ReportViewSetTests(TestCase):
+    def setUp(self):
+        self.user = CustomUser(
+            email="testuser@example.com",
+            password='newuserpasswd',
+            first_name='John',
+            last_name='Doe',
+            country='PL'
+        )
+        self.user.full_clean()
+        self.user.save()
+
+        self.project = Project(
+            name="Test Project",
+            start_date=datetime.datetime.now(),
+        )
+        self.project.full_clean()
+        self.project.save()
+
+        self.report = Report(
+            date=datetime.datetime.now().date(),
+            description='Some description',
+            author=self.user,
+            project=self.project,
+            work_hours=Decimal('8.00'),
+        )
+        self.report.full_clean()
+        self.report.save()
+
+    """
+    -----------
+    REPORT LIST
+    -----------
+    """
+    def test_report_list_view_should_display_users_report_list(self):
+        request = APIRequestFactory().get(path=reverse('report-list'))
+        request.user = self.user
+        response = ReportViewSet.as_view({'get': 'list'})(request)
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, self.report.description)
+
+    def test_report_list_view_should_not_be_accessible_for_unauthenticated_user(self):
+        request = APIRequestFactory().get(path=reverse('report-list'))
+        request.user = AnonymousUser()
+        response = ReportViewSet.as_view({'get': 'list'})(request)
+        self.assertEqual(response.status_code, 403)
+
+    def test_report_list_view_should_not_display_other_users_reports(self):
+        other_user=CustomUser(
+            email="otheruser@example.com",
+            password='otheruserpasswd',
+            first_name='Jane',
+            last_name='Doe',
+            country='PL',
+        )
+        other_user.full_clean()
+        other_user.save()
+
+        other_report = Report(
+            date=datetime.datetime.now().date(),
+            description='Some other description',
+            author=other_user,
+            project=self.project,
+            work_hours=Decimal('8.00'),
+        )
+        other_report.full_clean()
+        other_report.save()
+
+        request = APIRequestFactory().get(path=reverse('report-list'))
+        request.user = self.user
+        response = ReportViewSet.as_view({'get': 'list'})(request)
+        self.assertEqual(response.status_code, 200)
+        self.assertNotContains(response, other_report.description)
+
+    def test_that_report_list_view_should_add_new_report_on_post(self):
```

Comment:
The way this test is performed, I cannot verify if anything in the backend is happening.
I.e. I no new object is created in the base here, all I can see is the status code and template contents.

Is this an expected behaviour for views' unit tests? Is what I'm describing above an integration test?

If so, should rename those tests appropriately?

Reply:
Just check if the created object is in database after creation using views.

Reply:
Okay, so during my generic view research I found the expected codes for each operation.
https://www.django-rest-framework.org/api-guide/generic-views/#listmodelmixin

It's nice since those codes describe the exact behaviour I want to test, but now I have a different issue: I can fiddle with multiple HTTP requests here and get same result.
e.g.: If I were to change 'put' to either 'get' or 'post', the status code is always the same.

I just don't get those Rest viewsets :/

Reply:
So in REST you have some operations with specified url and method:
- `GET /objects/` one object details. Returns `200`. Database operation: `read`.
- `GET /objects/<id>/` objects list. Returns `200`. Database operation: `read`.
- `POST /objects/` object creation. Returns `201`. Database operation: `write`.
- `DELETE /objects/<id>/` object deletion. Returns `200`. Database operation: `delete`.
- `UPDATE /objects/<id>/` object partial update. Returns `200`. Database operation: `update`.
- `PUT /objects/<id>/` object full update. Returns `200`. Database operation: `update`.

You can distinct particular endpoints checking response and expected result in database. Status code is a confirmation that everything was done correctly.

Reply:
I just wonder how am I supposed to interpret action defined as {'put':'create'}. So 'put' is obviously HTTP request and 'create' refers to a viewset element (much like 'retrieve' and 'list' in other cases). But what does it declare here? Why if I were to replace 'put' with 'get' or 'post' it would always return a 201?

Reply:
I think it's DRF bug. Or you can check if method is allowed for specific view or not. This should not happen. You can add chore for that in Pivotal.